### PR TITLE
[ECO-5148] Implement “async room get” spec points

### PR DESCRIPTION
**Pull request by lawrence-forooghian**
**Note: This PR is based on top of #163; please review that one first.**

Implements spec points CHA-RC1f and CHA-RC1g.

## Summary by CodeRabbit

- **New Features**
	- Introduced a new error case for improved error handling in the Ably Chat SDK.
	- Enhanced room state management with a more structured approach to tracking room statuses.

- **Bug Fixes**
	- Improved error messaging for specific error cases to enhance user clarity.

- **Tests**
	- Expanded test coverage for room management functionalities, including asynchronous operations and error scenarios.
	- Added new helper classes and methods to facilitate testing of release operations.

- **Chores**
	- Updated mock classes to track method calls for better testing insights.

### CODE REVIEW

**Comment by lawrence-forooghian** [comment]
Tests are failing due to #169.

File: Sources/AblyChat/Rooms.swift [hunk 1]
```diff
@@ -21,8 +21,55 @@ internal actor DefaultRooms<RoomFactory: AblyChat.RoomFactory>: Rooms {
     private let logger: InternalLogger
     private let roomFactory: RoomFactory
 
-    /// The set of rooms, keyed by room ID.
-    private var rooms: [String: RoomFactory.Room] = [:]
+    /// All the state that a `DefaultRooms` instance might hold for a given room ID.
+    private enum RoomState {
+        /// There is no room map entry (see ``RoomMapEntry`` for meaning of this term) for this room ID, but a CHA-RC1g release operation is in progress.
+        case releaseOperationInProgress(releaseTask: Task<Void, Never>)
```
Comment (maratal) [inline Sources/AblyChat/Rooms.swift:27]: Why room can't be kept in the map until it fully released so you don't need additional enum and logic associated with it?

**Review by maratal — COMMENTED** [review]
Asked some questions

File: Sources/AblyChat/Rooms.swift [hunk 2]
```diff
@@ -32,39 +79,222 @@ internal actor DefaultRooms<RoomFactory: AblyChat.RoomFactory>: Rooms {
...
         }
 
-        // CHA-RC1d
-        rooms.removeValue(forKey: roomID)
+        switch roomState {
+        case let .releaseOperationInProgress(releaseTask):
+            // CHA-RC1g3
+            await waitForOperation(releaseTask, waitingOperationType: .release, waitedOperationType: .release)
+        case let .roomMapEntry(
+            .requestAwaitingRelease(
```
[the remark this answers is not in the document]
Comment (lawrence-forooghian) [reply Sources/AblyChat/Rooms.swift:272]: I’m not sure I understand. `releaseTask` and `failCreation` are values associated with the `.requestAwaitingRelease` enum case. Please could you clarify your question?

File: Sources/AblyChat/Rooms.swift (hunk 1 above)
Comment (lawrence-forooghian) [reply Sources/AblyChat/Rooms.swift:27]: The “room map” is a concept of the spec, and the spec is prescriptive about what room map manipulations should be performed, so I didn’t want to mess with that.

File: Sources/AblyChat/Rooms.swift (hunk 2 above)
[the remark this answers is not in the document]
Comment (maratal) [reply Sources/AblyChat/Rooms.swift:272]: uh oh, I was lost in this swift construct. So those are `let`s, I see now.

**Review by maratal — APPROVED** [review]
LGTM

**Pull request merged**